src/writer/agents.py

```diff
@@ -0,0 +1,383 @@
+from abc import ABC, abstractmethod
+import re
+from typing import Any, Dict, List, Optional, Union
+from enum import Enum
+import json
+
+class AgentOutput:
+    """Represents the output from an agent's execution"""
+    def __init__(self, response: str, reasoning: List[str], actions: List[Dict[str, Any]]):
+        self.response = response
+        self.reasoning = reasoning
+        self.actions = actions
+
+class Tool:
+    """Represents a tool that can be used by agents"""
+    def __init__(self, name: str, description: str, func: callable):
+        self.name = name
+        self.description = description
+        self.func = func
+        
+    def run(self, *args, **kwargs) -> Any:
+        return self.func(*args, **kwargs)
+
+class BaseAgent(ABC):
+    """Base class for all agents"""
+    def __init__(self, tools: Optional[List[Tool]] = None):
+        self.tools = tools or []
+        self.conversation_history = []
+        
+    @abstractmethod
+    def run(self, prompt: str) -> AgentOutput:
+        """Run the agent with the given prompt"""
+        pass
+
+    def add_tool(self, tool: Tool):
+        """Add a tool to the agent's toolset"""
+        self.tools.append(tool)
+
+    def _format_tool_descriptions(self) -> str:
+        """Format tool descriptions for prompt"""
+        if not self.tools:
+            return ""
+        return "Available tools:\n" + "\n".join(
+            f"- {tool.name}: {tool.description}" 
+            for tool in self.tools
+        )
+
+class ReActAgent(BaseAgent):
+    """
+    ReAct (Reasoning + Acting) agent that follows a structured thought process:
+    1. Reasoning about the current situation
+    2. Planning actions based on reasoning
+    3. Executing actions and observing results
+    4. Repeating until task completion
+    """
+    
+    def run(self, prompt: str) -> AgentOutput:
+        thoughts = []
+        actions = []
+        
+        # Add tools to system prompt if available
+        system_prompt = f"""Follow these steps to answer the user's question:
+1. Reason: Think about what needs to be done
+2. Action: Decide what action to take based on reasoning
+3. Observation: Consider the results
+4. Response: Provide final answer based on observations
+
+{self._format_tool_descriptions()}
+"""
+        # Initialize conversation with system prompt and user query
+        messages = [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": prompt}
+        ]
+        
+        # Add to conversation history
+        self.conversation_history.extend([
+            {"role": "user", "content": prompt}
+        ])
+        
+        response = None
+        max_steps = 5  # Prevent infinite loops
```

Comment:
Both `system prompt` and `max_steps` should not be hardcoded.
The prompt needs to be refined, I see next regex failing a lot:
```
match = re.search(r"Reason:(.+?)(?=Action:|Response:|$)", text, re.DOTALL)
```
I did not guess from the prompt what is the expected format, the model will struggle with this too.